src/serde/deserialize_utils.rs

```diff
@@ -0,0 +1,359 @@
+use std::ops::Rem;
+
+#[derive(Debug, PartialEq)]
+pub enum ReferenceParseError {
+    IntError(ParseIntError),
+    BigIntError(ParseBigIntError),
+    InvalidStringError(String),
+}
+
+impl fmt::Display for ReferenceParseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ReferenceParseError::IntError(error) => {
+                write!(f, "Int parsing error: ")?;
+                error.fmt(f)
+            }
+            ReferenceParseError::BigIntError(error) => {
+                write!(f, "BigInt parsing error: ")?;
+                error.fmt(f)
+            }
+            ReferenceParseError::InvalidStringError(error) => {
+                write!(f, "Invalid reference string error: ")?;
+                error.fmt(f)
+            }
+        }
+    }
+}
+
+// Checks if the hex string has an odd length.
+// If that is the case, prepends '0' to it.
+pub fn maybe_add_padding(mut hex: String) -> String {
+    if hex.len().rem(2) != 0 {
+        hex.insert(0, '0');
+        return hex;
+    }
+    hex
+}
+
+pub fn parse_dereference(value: &str) -> Result<ValueAddress, ReferenceParseError> {
+    let splitted: Vec<&str> = value.split(" + ").collect();
+
+    match splitted.len() {
+        1 => parse_dereference_no_offsets(splitted),
+        2 => parse_dereference_with_one_offset(splitted),
+        3 => parse_dereference_with_two_offsets(splitted),
+        _ => Err(ReferenceParseError::InvalidStringError(String::from(value))),
+    }
+}
+// parse string values of format `[cast(reg, *felt)]`
+fn parse_dereference_no_offsets(
+    splitted_value_str: Vec<&str>,
+) -> Result<ValueAddress, ReferenceParseError> {
+    let str_tmp: Vec<&str> = splitted_value_str[0].split(',').collect();
+
+    let register = match str_tmp[0].split('(').collect::<Vec<_>>()[1] {
+        "ap" => Some(Register::AP),
+        "fp" => Some(Register::FP),
+        _ => None,
+    };
+
+    Ok(ValueAddress {
+        register,
+        offset1: 0,
+        offset2: 0,
+        immediate: None,
+        dereference: true,
+    })
+}
+
+// parse string values of format `[cast(reg + offset1, *felt)]`
+fn parse_dereference_with_one_offset(
+    splitted_value_str: Vec<&str>,
+) -> Result<ValueAddress, ReferenceParseError> {
+    let register = match splitted_value_str[0].split('(').collect::<Vec<&str>>()[1] {
+        "ap" => Some(Register::AP),
+        "fp" => Some(Register::FP),
+        _ => None,
+    };
+
+    let mut offset1_str = splitted_value_str[1].split(',').collect::<Vec<_>>()[0].to_string();
+    offset1_str.retain(|c| !r#"()]"#.contains(c));
+
+    let offset1: i32 = match offset1_str.parse() {
+        Ok(offset1) => offset1,
+        // for the moment, references with values that overflow i32 are not important, they are just dummy references.
+        Err(e) => match e.kind() {
+            IntErrorKind::PosOverflow => 0,
+            _ => return Err(ReferenceParseError::IntError(e)),
+        },
```

Comment:
In fact, as long as they're offsets rather than values, they should remain in `i16` ranges, so this is correct code _under that assumption_.

Reply:
there are same cases where references (not used by hints, just in de deserializing process) have values of the form `cast(num, felt*)`, and `num` can be a very large integer, and the overflow happens. We are not interested in dealing with this values ATM so if the overflow occurs, returning 0 shoud be fine.